pkg/api/validate/quota.go

```diff
@@ -0,0 +1,93 @@
+package validate
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sort"
+
+	mgmtcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+)
+
+// checkQuota checks usage quotas vs. resources required by cluster before cluster creation
+func (dv *openShiftClusterDynamicValidator) validateQuotas(ctx context.Context, oc *api.OpenShiftCluster) error {
+	//prepare usage client
+	spp := &oc.Properties.ServicePrincipalProfile
+	r, err := azure.ParseResourceID(oc.ID)
+	if err != nil {
+		return err
+	}
+	conf := auth.NewClientCredentialsConfig(spp.ClientID, string(spp.ClientSecret), spp.TenantID)
+	authorizer, err := conf.Authorizer()
+	if err != nil {
+		return err
+	}
+	usageClient := compute.NewUsageClient(r.SubscriptionID, authorizer)
+	return dv.calculateAndCompareQuota(ctx, oc, usageClient)
+}
+
+func (dv *openShiftClusterDynamicValidator) addRequiredResources(requiredResources map[string]int, vmSize api.VMSize, count int) error {
+	requiredResources["virtualMachines"] += count
+	requiredResources["PremiumDiskCount"] += count
+	switch vmSize {
+	case api.VMSizeStandardD2sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 2)
+	case api.VMSizeStandardD4sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 4)
+	case api.VMSizeStandardD8sV3:
+		requiredResources["standardDSv3Family"] += count
+		requiredResources["cores"] += (count * 8)
+	default:
+		//will only happen if pkg/api verification allows new VMSizes
+		return fmt.Errorf("Unexpected node VMSize %s.", vmSize)
+	}
+	return nil
+}
+
+func (dv *openShiftClusterDynamicValidator) calculateAndCompareQuota(ctx context.Context, oc *api.OpenShiftCluster, uc compute.UsageClient) error {
+	requiredResources := map[string]int{}
+	dv.addRequiredResources(requiredResources, oc.Properties.MasterProfile.VMSize, 3)
+	//worker node resource calculation
+	for _, w := range oc.Properties.WorkerProfiles {
+		dv.addRequiredResources(requiredResources, w.VMSize, w.Count)
+	}
+
+	usages, err := uc.List(ctx, oc.Location)
+	if err != nil {
+		return err
+	}
+	//check requirements vs. usage
+	var resources []string
+	for r := range requiredResources {
+		resources = append(resources, r)
+	}
+	sort.Strings(resources)
+
+	for _, r := range resources {
+		var usage *mgmtcompute.Usage
```

Comment:
create a dummy usage with all the necessary fields (name, limit=0, currentvalue=0) populated to avoid needing the `usage == nil` case below.

Reply:
In that case the customer will get "Resource quota of X exceeded. Maximum allowed: 0, Current in use: 0, Additional requested: 1." 400 error, while the issue would be that the Azure Usage API didn't send resource usage. It might take a while to route a support ticket to the right place.

The two options I considered for when the Usage API doesn't send limits on one of the resources we check are:
1. As discussed in https://github.com/Azure/ARO-RP/pull/192#discussion_r381318499 server side error reflected in the logs
2. Ignore missing limit and proceed with cluster creation.

If we take option 2 the installation will fail when deploying the ARM template (assuming the limit was not returned by usage API, but exists) or continue (if there is no limit on a resource anymore).

Reply:
I think option 2 is OK, sure